alf/algorithms/ppg_algorithm.py

```diff
@@ -0,0 +1,391 @@
+                simultateously. env only needs to be provided to the root
+                Algorithm.
+            config (TrainerConfig): config for training. config only needs to be
+                provided to the algorithm which performs ``train_iter()`` by
+                itself.
+            encoding_network_ctor (Callable[..., Network]): Function to
+                construct the encoding network. The constructed network will be
+                called with ``forward(observation, state)``.
+            policy_optimizer (torch.optim.Optimizer): The optimizer for training
+                the policy phase of PPG.
+            aux_optimizer (torch.optim.Optimizer): The optimizer for training
+                the auxiliary phase of PPG.
+            epsilon_greedy (float): a floating value in [0,1], representing the
+                chance of action sampling instead of taking argmax. This can
+                help prevent a dead loop in some deterministic environment like
+                Breakout. Only used for evaluation. If None, its value is taken
+                from ``alf.get_config_value(TrainerConfig.epsilon_greedy)``. It
+                is used in ``predict_step()`` during evaluation.
+            debug_summaries (bool): True if debug summaries should be created.
+            name (str): Name of this algorithm.
+
+        """
+
+        dual_actor_value_network = DisjointPolicyValueNetwork(
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            encoding_network_ctor=encoding_network_ctor,
+            is_sharing_encoder=False)
+
+        super().__init__(
+            config=config,
+            env=env,
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            reward_spec=reward_spec,
+            predict_state_spec=dual_actor_value_network.state_spec,
+            train_state_spec=dual_actor_value_network.state_spec)
+
+        self._aux_phase_interval = aux_phase_interval
+        if epsilon_greedy is None:
+            epsilon_greedy = alf.get_config_value(
+                'TrainerConfig.epsilon_greedy')
+        self._predict_step_epsilon_greedy = epsilon_greedy
+
+        # The policy phase uses the main experience replayer with the PPO Loss
+        self._policy_phase = PPGPhaseContext(
+            main_algorithm=self,
+            network=dual_actor_value_network,
+            optimizer=policy_optimizer,
+            loss=PPOLoss(debug_summaries=debug_summaries),
+            exp_replayer=None)
+
+        # The auxiliary phase uses an extra experience replayer with the loss
+        # specific to the auxiliary update phase.
+        self._aux_phase = PPGPhaseContext(
+            main_algorithm=self,
+            network=dual_actor_value_network.copy(),
+            optimizer=aux_optimizer,
+            loss=PPGAuxPhaseLoss(),
+            exp_replayer=OnetimeExperienceReplayer())
```

Comment:
This will not work if PPGAlgorithm is used as a sub-algorithm. Currently, only the root algorithm automatically stores experience to replay buffer. You have to store experience to the replay buffer for aux phase by yourself at rollout_step().

I suggest you put PPG in Agent to test if the whole thing works.

Reply:
> This will not work if PPGAlgorithm is used as a sub-algorithm. Currently, only the root algorithm automatically stores experience to replay buffer. You have to store experience to the replay buffer for aux phase by yourself at rollout_step().

Yep I picked and remembered the part from last round of review already - trusted your suggestion. Haven't done this because I haven't fully understood why it does not work (yet). Will need to try this with `Agent` to see how it works.

> I suggest you put PPG in Agent to test if the whole thing works.

Good point. I will try this.

Reply:
Updated the design and now it works both standalone and under `Agent`. Tried to make this as elegant as possible.